### ISSUE TITLE
Will barcode scanning detection/speed be improved?

### ISSUE THREAD

**Issue by xGary** [googlesamples/android-vision#5]
Normally I don't like to post questions in issue trackers, but I think this will be of interest to a lot of developers using this API. So here it goes:

I know this API is new, but I am wondering if there will be improvements coming or will there just be fixes and maintenance?

The reason is I played around with the barcode scanning and while it is an improvement to existing free libraries like Zxing, paid options like [Scandit](https://play.google.com/store/apps/details?id=com.scandit.demoapp&hl=en) are still significantly faster. They seem to have a lot better low-light and blurry recognition which allows them to recognize the barcode a lot faster by not having to wait for the camera's autofocus. Unfortunately, Scandit's pricing is very steep and most developers either cannot afford it or decide to not spend the money on it. I think it would greatly improve all apps that use barcode reading if Google can offer a close or even a superior free option. And there is no reason why you guys at Google can't do that when you're already the very best at many things.

**Comment by juliangsf** [comment]
Thanks xGary
Yes - we want to improve this, and appreciate your feedback to help us do so. I believe you've seen Gericorp's suggested Autofocus workaround for now:
https://gist.github.com/Gericop/7de0b9fdd7a444e53b5a
Any other suggestions we should prioritize?  Presumably the faster the better, but is there a specific speed you need?
I'll also add a link shortly to collect feedback.

**Comment by xGary** [comment]
Thanks!

I have used the autofocus workaround, but relying on autofocus is really the thing that is holding the detection back. I think if you try out Scandit's demo scanner in the [Google Play Store](https://play.google.com/store/apps/details?id=com.scandit.demoapp&hl=en), you will quickly see what makes it superior. 

I think speed is definitely the first priority for most developers and their users. I think Scandit's ability to read blurry, unfocused barcodes is what makes it way faster. The Vision API is nearly instant at reading the barcode once the barcode is in focus, but waiting for the camera to focus will take a little bit of time especially on older devices. For example, I have an older phone (HTC One M7) that takes about ~2s to focus on something, once focused, the barcode gets read instantly, but it still took 2s because it relied on waiting for the camera to autofocus. When using Scandit's demo, it no longer nears to wait till the barcode is in focus so it is usually nearly instant or <300ms

**Comment by juliangsf** [comment]
Thanks very much xGary - very helpful. We understand your issue and will take a look.

**Comment by Gericop** [comment]
@xGary, you may want to try experimenting with the different focus modes. For example, continuous picture (*FOCUS_MODE_CONTINUOUS_PICTURE*) is more aggressive than the video one (*FOCUS_MODE_CONTINUOUS_VIDEO*) so the detectors might have to wait more until the focus is set. Meanwhile, the video one is smoother, thus the detector might have some chance to detect the barcodes during the focusing process.
*Disclaimer: this is only an opinion, I haven't tested it.*

One more thing you might find useful: try to set the focus only once (for example, trigger it manually) when you think (or the software thinks, create your algorithm for this) that the barcode is at an appropriate distance. This way, all the closer barcodes will work almost instantly since no refocusing is needed, although if you have things in the far, you might have to focus on them again. So I think the easiest way would be to let the user focus on the first barcode then just use the same focus to read more barcodes. If they want to scan farther objects, they will have to refocus, but the closer ones will work instantly.

**Comment by mindshifter** [comment]
this is my fix with Sensor:

@Override
    public void onSensorChanged(SensorEvent event) {
        if (mCameraSource != null) {
            float x = event.values[0];
            float y = event.values[1];
            float z = event.values[2];
            Log.e(TAG, "float x = event.values[0] = " + String.valueOf(x));
            if (!mInitialized) {
                mLastX = x;
                mLastY = y;
                mLastZ = z;
                mInitialized = true;
            }
            float deltaX = Math.abs(mLastX - x);
            float deltaY = Math.abs(mLastY - y);
            float deltaZ = Math.abs(mLastZ - z);
            if (deltaX > .5 || deltaY > .5 || deltaZ > .5) {
                mCameraSource.autoFocus(myAutoFocusCallback);
            }
            mLastX = x;
            mLastY = y;
            mLastZ = z;
        }
    }

**Comment by danielgindi** [comment]
Actually just comparing the speed of detection to the built in detection technology in iOS - shows a great difference.
In iOS it's much much faster - and very accurate.
So if I show an assistive barcode tracking graphic over the preview layout, the users feel a great difference as in iOS it is much smoother (updates around 10 times per second), and in Android it updates about 2-3 times a second. On relatively new and powerful devices.
On iOS it's very easy to move the graphic to it's updated location using a UIView animation, but on Android it gets tougher - at least with the sample code - and the locations are jerky to begin with.

So an overall improvement would be great.

**Comment by i-schuetz** [comment]
Same problem for the text reader! I want to show overlays on top of the (filtered) text but with the autofocus delays this isn't usable.

**Comment by satyaphanindra** [comment]
I agree with @danielgindi. I have also observed in my project (where the camera get to capture a qrcode which is on the phone screen when placed in front of a mirror) capturing a qrcode is fast when camera is near to the visible qrcode. 
Though I don't like comparing, just to mention that the iOS device is just faster even if the camera is a not very near to the qrcode and also captures even if qrcode is small.

**Comment by LaxmikanthMadhyastha** [comment]
Did anyone find a workaround for this or any technique where the speed can be improved?
